// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

syntax = "proto3";

package monorail.v1;

import "google_proto/google/api/field_behavior.proto";
import "google_proto/google/api/resource.proto";
import "google_proto/google/api/annotations.proto";
import "api/v1/api_proto/project_objects.proto";


// Projects service includes all methods needed for managing Projects.
service Projects {
  // Returns all templates for specified project.
  //
  // Raises:
  //   NOT_FOUND if the requested parent project is not found.
  //   INVALID_ARGUMENT if the given `parent` is not valid.
  rpc ListIssueTemplates (ListIssueTemplatesRequest) returns (ListIssueTemplatesResponse) {
    option (google.api.http) = {
      post: "/prpc/monorail.v1.Projects/ListIssueTemplates"
      body: "*"
    };
  }

  // Returns all project specific configurations needed for the SPA client.
  rpc FetchProjectEnvironment (FetchProjectEnvironmentRequest) returns (FetchProjectEnvironmentResponse) {
    option (google.api.http) = {
      post: "/prpc/monorail.v1.Projects/FetchProjectEnvironment"
      body: "*"
    };
  }
}


// Request message for ListIssueTemplates
// Next available tag: 4
message ListIssueTemplatesRequest {
  // The name of the project these templates belong to.
  string parent = 1 [
    (google.api.resource_reference) = {type: "api.crbug.com/Project"},
    (google.api.field_behavior) = REQUIRED ];
  // The maximum number of items to return. The service may return fewer than
  // this value.
  int32 page_size = 2;
  // A page token, received from a previous `ListIssueTemplatesRequest` call.
  // Provide this to retrieve the subsequent page.
  // When paginating, all other parameters provided to
  // `ListIssueTemplatesRequest` must match the call that provided the token.
  string page_token = 3;
}


// Response message for ListIssueTemplates
// Next available tag: 3
message ListIssueTemplatesResponse {
  repeated IssueTemplate templates = 1;
  // A token, which can be sent as `page_token` to retrieve the next page.
  // If this field is omitted, there are no subsequent pages.
  string next_page_token = 2;
}


// Request message for FetchProjectEnvironment
// Next available tag: 2
message FetchProjectEnvironmentRequest {
  // The name of the project these config environments belong to.
  string parent = 1 [
    (google.api.resource_reference) = {type: "api.crbug.com/Project"},
    (google.api.field_behavior) = REQUIRED ];
}


// Response message for FetchProjectEnvironment
// Next available tag: 21
message FetchProjectEnvironmentResponse {
  // List of well known statuses that belong to this project.
  repeated StatusDef status_defs = 1;
  // List of well known labels that belong to this project.
  repeated LabelDef label_defs = 2;
  // List of components that belong to this project.
  repeated ComponentDef component_defs = 3;
  // List of custom fields that belong to this project.
  repeated FieldDef field_defs = 4;
  // List of approval fields that belong to this project.
  repeated ApprovalDef approval_defs = 5;
  // If an issue's status is one of these, show "Merge with:".
  repeated StatusDef statuses_offer_merge = 6;
  // An issue may only have one label that begins with these prefixes
  // E.g. priority, since no issue can be both Priority-High and Priority-Low.
  repeated string exclusive_label_prefixes = 7;
  // Saved search queries that admins defined for this project.
  repeated ProjectSavedQuery saved_queries = 8;
  // Default search query for this project, only visible for project members.
  string default_query = 9;
  // Default sort specification for this project.
  string default_sort = 10;
  // Default list columns for this project.
  string default_col_spec = 11;
  // Default column dimension in grid view for this project.
  string default_x_attr = 12;
  // Default row dimension in grid view for this project.
  string default_y_attr = 13;
  // URL pointing to this project's logo image
  string project_thumbnail_url = 14;
  // URL format to browse source code revisions for this project.
  string revision_url_format = 15;
  // A project's custom URL for the "New issue" link, only if specified.
  string custom_issue_entry_url = 16;
  // Default template used for issue entry for members of this project.
  IssueTemplate default_template_for_members = 17;
  // Default template used for issue entry for non-members of this project.
  IssueTemplate default_template_for_non_members = 18;

  // It's debatable whether this is needed. For completeness of a resource,
  // this should be included, but since this isn't a "standard" resource and
  // the client already knows the project name it could be omitted.
  // I can go either way
  // Display name of this project, ie Monorail or Chromium
  string project_name = 19;
  // Short description of this project. Shown as tool-tip on project selector.
  string project_summary = 20;
}